src/dataflow/src/render/upsert.rs

```diff
@@ -136,77 +111,104 @@ where
                         time,
                     ) in vector.drain(..)
                     {
-                        let entry = values
-                            .entry(cap.delayed(&time))
-                            .or_insert_with(HashMap::new)
+                        if key.is_empty() {
+                            error!("{}", "Encountered empty key");
+                            continue;
+                        }
+                        let entry = to_send
+                            .entry(time)
+                            .or_insert_with(|| (cap.delayed(&time), HashMap::new()))
+                            .1
                             .entry(key)
                             .or_insert_with(Default::default);
 
+                        let new_entry = SourceData {
+                            value: new_value,
+                            position: new_position,
+                            upstream_time_millis: new_upstream_time_millis,
+                        };
+
+                        // if the time is equal, toss out the row with the
+                        // lower offset
                         if let Some(new_offset) = new_position {
```

Comment:
What is the `else` case for this? We didn't get an offset for some reason? Should we log an error?

Reply:
Yeah, the else case would be not getting an offset for some reason.

I believe the `else` case is currently considered unreachable, but since we prefer to not panic, I guess we should log an error in the event we hit the code that should be unreachable.